# Equipo 1
## Grupo: 001 (Lunes 7:00 - 10:00 pm)
### Integrantes:
+ 1848655 - Rangel Vallejo Isabel Alejandra 
+ 1851167 - Aguilar Sánchez José Antonio
+ 1851124 - Cerecero Peña Angela Scarlett
+ 1851528 - Romero Banda José Manuel
+ 1853849 - García Gutiérrez Alfonso de Jesús 

# CO2 Emission by Vehicles

**1) Título de la base de datos**  
a) Nombre con el que se encuentra en la página: **"CO2 Emission by Vehicles"**  
b) Url de la página: [CO2 Emission](https://www.kaggle.com/debajyotipodder/co2-emission-by-vehicles) 

**2) Introducción general**

<div style="text-align: justify">El CO2 (dióxido de carbono) es un compuesto de carbono y oxígeno que existe como gas incoloro en condiciones de temperatura y presión estándar. Es uno de los gases que conforman al efecto invernadero, las principales consecuencias del efecto invernadero son: Aumento de la temperatura media de la Tierra, reducción de la superficie de glaciares y, como consecuencia, elevación del nivel del agua de los mares y océanos.
<div style="text-align: justify">
Las emisiones excesivas de este gas son una de las principales causas del calentamiento global. Un problema causado por la actividad humana y agravado por la larga pervivencia del CO2 en la atmósfera. Los automóviles particulares generan el 18% de las emisiones de CO2.      
<div style="text-align: justify">
Una herramienta capaz de cuantificar los datos relevantes a las emisiones de CO2 producidas por automóviles ayudaría en gran medida a la humanidad a poder combatir este fenómeno que está agraviando cada vez más la situación del mundo. Al difundir las causas, características o patrones de las emisiones de este gas se estaría contribuyendo a disminuir su emisión. </div>

**3) Descripción de los datos**

Los datos que se obtuvieron se recopilaron a partir de una tabla que señala las distintas características que pudieran contribuir a las emisiones de CO2 hechas por automóviles. La tabla muestra la información de manera completa.

La tabla consta de 12 columnas, estas son: 
+ **Make** - Marca del vehículo (ejemplo: ACURA, ASTON MARTIN, etc.). 
+ **Modelo** - Modelo del automóvil (ejemplo: ILX, V8 VANTAGE, etc.). 
+ **Vehicle Class** - Clase del vehículo (ejemplo: COMPACT, MINIVAN, etc.). 
+ **Engine Size(L)** - Tamaño del motor del automóvil, viene representado en litros e indican la cantidad de gasolina que utiliza un coche cuando sus cilindros están llenos.
+ **Cylinders** - Cantidad de cilindros que tiene el vehículo. 
+ **Transmission** - Tipo de transmisión del automóvil. Se clasifica en:
    + A = Automático
    + AM = Automated manual
    + AS = Automatic with select shift
    + AV = Continuously variable
    + M = Manual
+ **Fuel Type** - Tipo de gasolina que usa el vehículo. Las cuales son:
    + X = Gasolina Regular 
    + Z = Gasolina Premium
    + D = Diesel
    + E = Etanol (E85)
    + N =Gas natural 
    + Número de marchas del automóvil:  de 3 a 10.
+ **Fuel Consumption City (L/100 km)** -  Consumo de combustible en la ciudad; se muestra en litros por cada 100 kilómetros (L / 100 km).
+ **Fuel Consumption Hwy (L/100 km)** - Consumo de combustible en la autopista; se muestra en litros por cada 100 kilómetros (L / 100 km).
+ **Fuel Consumption Comb (L/100 km)** - Consumo de combustible combinada (55% en la ciudad y 45% en la carretera); se muestra en litros por cada 100 kilómetros (L / 100 km).
+ **Fuel Consumption Comb (mpg)** - Consumo de combustible combinada (55% en la ciudad y 45% en la carretera); se muestra en millas por galón (mpg)
+ **CO2 Emissions(g/km)** - Emisiones de dióxido de carbono del tubo de escape (en gramos por kilómetro) de una conducción combinada en ciudad y carretera.

**4) Justificación del uso de datos**  

<div style="text-align: justify">Como se mencionó anteriormente, el calentamiento global es detonado por las emisiones de CO2 provocadas por las actividades humanas. Se sabe que los automóviles son responsables de esto en un 18%. Teniendo en cuenta esto último, esta base es un complemento espectacular para combatir dicho problema debido a que se puede identificar las cualidades, patrones o características que un vehículo debe poseer para emitir cierto nivel de CO2.</div>
<div style="text-align: justify">
De la misma manera, se podría estudiar si el rendimiento que un automóvil tiene (en ciudad o carretera) contribuye a este problema. Juntando todo esto, resulta sencillo resaltar que la motivación de trabajar con estos datos es la gama tan amplia que se tiene al momento de estudiar el fenómeno en cuestión. Así mismo, se obtiene un beneficio al conocer los resultados arrojados por la investigación; por ejemplo, al saber qué compañías fabrican los vehículos que contaminan más que otros, se puede lograr una conciencia al momento de adquirir un automóvil y, de esta manera, ayudar al ambiente.</div>

**5) Objetivo final**

Objetivo principal:  
+ Predecir las emisiones de CO2 (g/km) de los automóviles

Objetivos secundarios:
+ Establecer si el tipo de gasolina influye en las emisiones producidas
+ Determinar las armadoras que tienen menor índice de emisiones

**6) Mencionar preguntas de interés**
+ ¿Qué clases de vehículo representan un mayor riesgo al ambiente?
+ ¿La transmisión del automóvil hace una diferencia al momento de emitir CO2?

**7) Parte de programación**

## Limpieza de datos de la base principal: "CO2 Emission by Vehicles"

Empezaremos con la importación de las librerías que se utilizarán

In [1]:
import pandas as pd
import numpy as np

Continuaremos con la lectura de nuestra base de datos que previamente fue descargada de la página web mencionada anteriormente

In [2]:
baseCO2 = pd.read_csv('CO2.csv')

La base de datos está guardada en la variable "baseCO2"

## Visualización

Una vez que la base se ha leído correctamente, se procede a analizar la estructura de la misma de manera completa

In [3]:
baseCO2

Make        Model   Vehicle Class  Engine Size(L)  Cylinders  \
0     ACURA          ILX         COMPACT             2.0          4   
1     ACURA          ILX         COMPACT             2.4          4   
2     ACURA   ILX HYBRID         COMPACT             1.5          4   
3     ACURA      MDX 4WD     SUV - SMALL             3.5          6   
4     ACURA      RDX AWD     SUV - SMALL             3.5          6   
...     ...          ...             ...             ...        ...   
7380  VOLVO  XC40 T5 AWD     SUV - SMALL             2.0          4   
7381  VOLVO  XC60 T5 AWD     SUV - SMALL             2.0          4   
7382  VOLVO  XC60 T6 AWD     SUV - SMALL             2.0          4   
7383  VOLVO  XC90 T5 AWD  SUV - STANDARD             2.0          4   
7384  VOLVO  XC90 T6 AWD  SUV - STANDARD             2.0          4   

     Transmission Fuel Type  Fuel Consumption City (L/100 km)  \
0             AS5         Z                               9.9   
1              M6         Z                              11.2   
2             AV7         Z                               6.0   
3             AS6         Z                              12.7   
4             AS6         Z                              12.1   
...           ...       ...                               ...   
7380          AS8         Z                              10.7   
7381          AS8         Z                              11.2   
7382          AS8         Z                              11.7   
7383          AS8         Z                              11.2   
7384          AS8         Z                              12.2   

      Fuel Consumption Hwy (L/100 km)  Fuel Consumption Comb (L/100 km)  \
0                                 6.7                               8.5   
1                                 7.7                               9.6   
2                                 5.8                               5.9   
3                                 9.1                              11.1   
4                                 8.7                              10.6   
...                               ...                               ...   
7380                              7.7                               9.4   
7381                              8.3                               9.9   
7382                              8.6                              10.3   
7383                              8.3                               9.9   
7384                              8.7                              10.7   

      Fuel Consumption Comb (mpg)  CO2 Emissions(g/km)  
0                              33                  196  
1                              29                  221  
2                              48                  136  
3                              25                  255  
4                              27                  244  
...                           ...                  ...  
7380                           30                  219  
7381                           29                  232  
7382                           27                  240  
7383                           29                  232  
7384                           26                  248  

[7385 rows x 12 columns]

Se puede observar que la base ha cargado y muestra algunos de los elementos que contiene

En ocasiones se desea ver algunos pocos elementos de la base y los nombres de la columna para empezar a estar familizarizados con la base, por lo que es muy útil emplear la función siguiente

In [4]:
baseCO2.head(3)

Make       Model Vehicle Class  Engine Size(L)  Cylinders Transmission  \
0  ACURA         ILX       COMPACT             2.0          4          AS5   
1  ACURA         ILX       COMPACT             2.4          4           M6   
2  ACURA  ILX HYBRID       COMPACT             1.5          4          AV7   

  Fuel Type  Fuel Consumption City (L/100 km)  \
0         Z                               9.9   
1         Z                              11.2   
2         Z                               6.0   

   Fuel Consumption Hwy (L/100 km)  Fuel Consumption Comb (L/100 km)  \
0                              6.7                               8.5   
1                              7.7                               9.6   
2                              5.8                               5.9   

   Fuel Consumption Comb (mpg)  CO2 Emissions(g/km)  
0                           33                  196  
1                           29                  221  
2                           48                  136

Así, podemos tener una idea de cómo está compuesta nuestra base

<div style="text-align: justify">Para tener organizada nuestra base, podemos ordenarla por alguna característica que nos sea relevante. En este caso, optaremos por ordenarla de acuerdo a la armadora, la cual es la columna "Make". Haciendo esto lograremos ordenar nuestros datos de A-Z.</div>

In [5]:
baseCO2 = baseCO2.sort_values('Make')
baseCO2

Make             Model   Vehicle Class  Engine Size(L)  Cylinders  \
0     ACURA               ILX         COMPACT             2.0          4   
4373  ACURA  MDX SH-AWD ELITE     SUV - SMALL             3.5          6   
4372  ACURA        MDX SH-AWD     SUV - SMALL             3.5          6   
4371  ACURA               ILX         COMPACT             2.4          4   
5454  ACURA               ILX         COMPACT             2.4          4   
...     ...               ...             ...             ...        ...   
5442  VOLVO        S90 T6 AWD        MID-SIZE             2.0          4   
5441  VOLVO        S90 T5 AWD        MID-SIZE             2.0          4   
5440  VOLVO     S60 CC T5 AWD         COMPACT             2.0          4   
5438  VOLVO        S60 T6 AWD         COMPACT             2.0          4   
7384  VOLVO       XC90 T6 AWD  SUV - STANDARD             2.0          4   

     Transmission Fuel Type  Fuel Consumption City (L/100 km)  \
0             AS5         Z                               9.9   
4373          AS9         Z                              12.2   
4372          AS9         Z                              12.6   
4371          AM8         Z                               9.4   
5454          AM8         Z                               9.9   
...           ...       ...                               ...   
5442          AS8         Z                              10.6   
5441          AS8         Z                              10.1   
5440          AS8         X                              10.8   
5438          AS8         Z                              10.9   
7384          AS8         Z                              12.2   

      Fuel Consumption Hwy (L/100 km)  Fuel Consumption Comb (L/100 km)  \
0                                 6.7                               8.5   
4373                              9.0                              10.7   
4372                              9.0                              11.0   
4371                              6.8                               8.2   
5454                              7.0                               8.6   
...                               ...                               ...   
5442                              7.6                               9.3   
5441                              7.4                               8.8   
5440                              7.8                               9.4   
5438                              7.7                               9.4   
7384                              8.7                              10.7   

      Fuel Consumption Comb (mpg)  CO2 Emissions(g/km)  
0                              33                  196  
4373                           26                  251  
4372                           26                  259  
4371                           34                  192  
5454                           33                  199  
...                           ...                  ...  
5442                           30                  216  
5441                           32                  206  
5440                           30                  220  
5438                           30                  221  
7384                           26                  248  

[7385 rows x 12 columns]

<div style="text-align: justify">A simple vista pareciera haber quedado igual que como se tenía, sin embargo, no podemos estar seguros de eso, ya que no mostraba la base de manera completa. Haciendo esto nos aseguramos de que esté organizada.</div>

## Características de la base de datos

Con el objetivo de hacer un análisis profundo sobre la base, se estudiarán las características que esta tiene.  
Comenzaremos con el número de filas y columnas

In [6]:
baseCO2.shape

(7385, 12)

Tiene 7385 filas y 12 columnas. Se puede interpretar como una base de datos amplia, ya que tiene muchos elementos que la conforman

<div style="text-align: justify">Ahora que sabemos la cantidad de filas y columnas que tiene nuestra base, podemos estudiar las características de las columnas a fondo, comenzando por el nombre de las mismas</div>

In [7]:
baseCO2.columns

Index(['Make', 'Model', 'Vehicle Class', 'Engine Size(L)', 'Cylinders',
       'Transmission', 'Fuel Type', 'Fuel Consumption City (L/100 km)',
       'Fuel Consumption Hwy (L/100 km)', 'Fuel Consumption Comb (L/100 km)',
       'Fuel Consumption Comb (mpg)', 'CO2 Emissions(g/km)'],
      dtype='object')

Los nombres de las columnas son los elementos que se han impreso. Podemos notar a simple vista que estos elementos están en inglés.

<div style="text-align: justify">Algunos de estos nombres no son lo suficientemente claros para hacer referencia a los elementos que contiene cada columna, de la misma manera, algunos nombres tienen espacios entre palabras, lo que podría generar un problema al momento de querer referenciar elementos que pertenezcan a dicha columna.</div>

Por esto último, se opta por cambiar los nombres de algunas columnas, con el fin de facilitar el manejo de los datos

In [8]:
# se crea nueva base con las modificaciones de los nombres de columnas

base_nueva = baseCO2.rename(columns = {'Vehicle Class':'Vehicle_Class','Engine Size(L)':'Engine_Size(L)',
                                       'Fuel Type':'Fuel_Type','Fuel Consumption City (L/100 km)':'Fuel_Consumption_City(L/100 km)',
                                       'Fuel Consumption Hwy (L/100 km)':'Fuel_Consumption_Hwy(L/100 km)',
                                        'Fuel Consumption Comb (mpg)':'Fuel_Consumption_Comb(mpg)',
                                       'CO2 Emissions(g/km)':'CO2_Emissions(g/km)'})

Corroboramos que el cambio se haya realizado:

In [9]:
base_nueva.columns

Index(['Make', 'Model', 'Vehicle_Class', 'Engine_Size(L)', 'Cylinders',
       'Transmission', 'Fuel_Type', 'Fuel_Consumption_City(L/100 km)',
       'Fuel_Consumption_Hwy(L/100 km)', 'Fuel Consumption Comb (L/100 km)',
       'Fuel_Consumption_Comb(mpg)', 'CO2_Emissions(g/km)'],
      dtype='object')

Vemos que todos nuestros cambios se han hecho correctamente

Otro problema que se pudiera presentar es en los tipos de datos. Necesitamos ver que nuestros datos estén guardados correctamente. Es decir, si son números, que sean valores numéricos y no de texto, entre otros. Para esto utilizamos la función dtypes

In [10]:
base_nueva.dtypes

Make                                 object
Model                                object
Vehicle_Class                        object
Engine_Size(L)                      float64
Cylinders                             int64
Transmission                         object
Fuel_Type                            object
Fuel_Consumption_City(L/100 km)     float64
Fuel_Consumption_Hwy(L/100 km)      float64
Fuel Consumption Comb (L/100 km)    float64
Fuel_Consumption_Comb(mpg)            int64
CO2_Emissions(g/km)                   int64
dtype: object

<div style="text-align: justify">Vemos que el nombre de las armadoras, los modelos, las clases de vehículos, transmisiones y el tipo de gasolina son de tipo objeto, lo cual es bueno, ya que estos elementos son string que nos ayudan a categorizar, los cuales no presentan problema. Ahora bien, el tamaño del motor y los consumos de gasolina están dados como flotantes, lo que de nueva cuenta no representa inconveniente alguno. Por último, los cilindros y las emisiones de C02 son valores enteros, lo cual también está bien para nuestro análisis.</div>

Podemos concluir finalmente que nuestros datos no presentan problemas con su tipo de datos.

Otra característica importante a revisar es la composición de las columnas. Contestar las preguntas: ¿Qué elementos tiene?, ¿Cuántos elementos distintos hay? son cosas que se pueden hacer

In [11]:
# Revisando los elementos de una columna
base_nueva['Make']

0       ACURA
4373    ACURA
4372    ACURA
4371    ACURA
5454    ACURA
        ...  
5442    VOLVO
5441    VOLVO
5440    VOLVO
5438    VOLVO
7384    VOLVO
Name: Make, Length: 7385, dtype: object

Podemos oobtener todos los elementos de la columna "Make", sin embargo, no vemos cuántos distintos hay, por esto se usa la siguiente función

In [12]:
base_nueva['Make'].unique()

array(['ACURA', 'ALFA ROMEO', 'ASTON MARTIN', 'AUDI', 'BENTLEY', 'BMW',
       'BUGATTI', 'BUICK', 'CADILLAC', 'CHEVROLET', 'CHRYSLER', 'DODGE',
       'FIAT', 'FORD', 'GENESIS', 'GMC', 'HONDA', 'HYUNDAI', 'INFINITI',
       'JAGUAR', 'JEEP', 'KIA', 'LAMBORGHINI', 'LAND ROVER', 'LEXUS',
       'LINCOLN', 'MASERATI', 'MAZDA', 'MERCEDES-BENZ', 'MINI',
       'MITSUBISHI', 'NISSAN', 'PORSCHE', 'RAM', 'ROLLS-ROYCE', 'SCION',
       'SMART', 'SRT', 'SUBARU', 'TOYOTA', 'VOLKSWAGEN', 'VOLVO'],
      dtype=object)

Con esto obtenemos todos los valores distintos que se encuentran en la columna "Make"

Resulta interesante hacerlo con las columnas "Model" y "Vehicle_Class"

In [13]:
base_nueva['Model'].unique()

array(['ILX', 'MDX SH-AWD ELITE', 'MDX SH-AWD', ..., 'S60 T6',
       'XC70 T5 AWD', 'V90 CC T5 AWD'], dtype=object)

In [14]:
base_nueva['Vehicle_Class'].unique()

array(['COMPACT', 'SUV - SMALL', 'MID-SIZE', 'TWO-SEATER', 'MINICOMPACT',
       'SUBCOMPACT', 'FULL-SIZE', 'STATION WAGON - SMALL',
       'SUV - STANDARD', 'PICKUP TRUCK - SMALL',
       'SPECIAL PURPOSE VEHICLE', 'PICKUP TRUCK - STANDARD',
       'VAN - PASSENGER', 'VAN - CARGO', 'MINIVAN',
       'STATION WAGON - MID-SIZE'], dtype=object)

## Valores nulos

<div style="text-align: justify">Continuando con el análisis de los datos, debemos observar a fondo los elementos que componen al dataframe. En vista de que tenemos un dataframe muy grande, resulta complicado visualizar toda la base en una sola exhibición, por lo que el estudio se realizará solo para encontrar valores nulos.</div>

Para esto se utiliza la siguiente función que retorna 'True' en la columna si esta contiene algún elemento nulo, o bien, 'False' si no contiene ninguno

In [15]:
base_nueva.isnull().any()

Make                                False
Model                               False
Vehicle_Class                       False
Engine_Size(L)                      False
Cylinders                           False
Transmission                        False
Fuel_Type                           False
Fuel_Consumption_City(L/100 km)     False
Fuel_Consumption_Hwy(L/100 km)      False
Fuel Consumption Comb (L/100 km)    False
Fuel_Consumption_Comb(mpg)          False
CO2_Emissions(g/km)                 False
dtype: bool

Con este resultado podemos decir que nuestra base de datos no contiene valores nulos. Para estar seguros, podemos revisar otra función que retorna el número de elementos nulos que contiene cada columna

In [16]:
base_nueva.isnull().sum()

Make                                0
Model                               0
Vehicle_Class                       0
Engine_Size(L)                      0
Cylinders                           0
Transmission                        0
Fuel_Type                           0
Fuel_Consumption_City(L/100 km)     0
Fuel_Consumption_Hwy(L/100 km)      0
Fuel Consumption Comb (L/100 km)    0
Fuel_Consumption_Comb(mpg)          0
CO2_Emissions(g/km)                 0
dtype: int64

Dado que obtuvimos 0 en todas las columnas, podemos afirmar que nuestra base no contiene elementos nulos

## Eliminación de columnas

Antes de eliminar columnas del dataframe, debemos analizar los objetivos que se tienen en mente al trabajar con esta base para no eliminar datos que se pudieran requerir

**Recordando...**  
Objetivo principal:  
+ Predecir las emisiones de CO2 (g/km) de los automóviles

Objetivos secundarios y preguntas de interés:
+ Establecer si el tipo de gasolina utilizada influye en las emisiones producidas
+ Determinar las armadoras que tienen menor índice de emisiones
+ ¿Qué clases de vehículo representan un mayor riesgo al ambiente?
+ ¿La transmisión del automóvil hace una diferencia al momento de emitir CO2?

**Recordando las columnas...**

In [17]:
base_nueva.columns

Index(['Make', 'Model', 'Vehicle_Class', 'Engine_Size(L)', 'Cylinders',
       'Transmission', 'Fuel_Type', 'Fuel_Consumption_City(L/100 km)',
       'Fuel_Consumption_Hwy(L/100 km)', 'Fuel Consumption Comb (L/100 km)',
       'Fuel_Consumption_Comb(mpg)', 'CO2_Emissions(g/km)'],
      dtype='object')

<div style="text-align: justify">Podemos notar de primera mano que, los consumos de gasolina, el tamaño del motor, los cilindros y las emisiones de CO2 se podrían utilizar para predecir las emisiones de CO2. Así como ocupamos a las armadoras, las clases de vehículo, el tipo de gasolina que se usa y la transmisión para algunas preguntas de interés y objetivos secundarios. Por lo que solo resta el modelo de los automóviles que es el que, a primera impresión, no se ha utilizado.</div>

<div style="text-align: justify">Sin embargo, esto no quiere decir que no se vaya a emplear en el trabajo. Como sabemos, el vehículo es el responsable de emitir el CO2, por lo que va relacionado directamente con este. Así mismo, podría emplearse en alguna gráfica que indice los modelos de vehículos con mayor emisión de CO2, entre otras posibles alternativas.</div>

Tomando todo esto en cuenta, se conluye que no se eliminará ninguna columna dado que en algún punto del trabajo podrían llegar a emplearse. Y, como sabemos, la información no se podría recuperar en caso de ser eliminada.

## Guardar un nuevo archivo

Para finalizar, se puede crear un nuevo archivo con todos los cambios que se han realizado, para que, cuando se vuelva a necesitar la base, no se tengan que realizar todos los cambios nuevamente.

In [18]:
base_nueva.to_csv('Emision_CO2_automoviles.csv', index=False, sep = ',')
# se crea un archivo csv separado o delimitado por comas

**----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------**

# MOVIES DATASET FOR FEATURE EXTRACION, PREDICTION

**1) Título de la base de datos**  
a) Nombre con el que se encuentra en la página: **"MOVIES DATASET FOR FEATURE EXTRACION ,PREDICTION"**  
b) Url de la página: [MOVIES DATASET](https://www.kaggle.com/bharatnatrayn/movies-dataset-for-feature-extracion-prediction) 

**2) Introducción general**

<div style="text-align: justify">La base de datos contiene información de distintas series y películas que se pueden encontrar en distintas plataformas como Netflix y Prime Video, estas aplicaciones cuentan con un catálogo muy variado de películas, series y documentales al que puedes acceder pagando una suscripción mensual o anual, a lo largo de los años se han hecho más reconocidos estos servicios de streaming y entretenimiento ya que la gente los empezó a consumir más por su gran contenido diverso con el que nunca te aburres y que, además, lo puedes ver y repetir las veces que quieras a diferencia de los canales de televisión.  
<div style="text-align: justify">
Una herramienta para explorar todos los datos que nos brinda una serie o película de cualquier plataforma serviría mucho a la hora de tomar decisiones que se deben hacer para la mejor administración de la empresa, ya que como se sabe, el contenido publicado en las aplicaciones se va actualizando y cambiando con el tiempo.</div>

**3) Descripción de los datos**

Los datos que se obtuvieron se encuentran en una tabla que muestra distintos aspectos de las series y películas.

La tabla consta de 9 columnas, estas son: 
+ **MOVIES** - Nombre de la serie o película. Compuesta en una cadena de texto.
+ **YEAR** - Año en el cual empezó su transmisión hasta la fecha que terminó en caso de series y en caso de películas el año en que se estrenó. Compuesta en cadena de texto.
+ **GENRE** - Género al que pertenece el programa o película. Cadena de texto.
+ **RATING** - Rating que tiene la serie o película. Número flotante.
+ **ONE-LINE** - Muestra un párrafo con una breve descripción de la serie/película.
+ **STARS** - Aparece el casting que participa en la grabación.
+ **VOTES** - Calificación de los votos que han dado las personas.
+ **RunTime** - Muestra el número entero de minutos que dura la película o serie.
+ **Gross** - Cantidad de dinero total en millones de dólares que se ganó con esta grabación.

**4) Justificación del uso de datos**  

Al analizar las variables y los datos proporcionados, nos llamó la atención las siguientes características:  
+ **La gran cantidad de datos:** al momento de examinar la base observamos que la cantidad de películas/series que se nos proporcionaba es de 9,999, entonces lo podemos tomar como que existe una gran variedad entre los datos.
+ **La variables:** al observarlas pudimos deducir que iban a ser fácil manejarlas y por lo tanto trabajar fácilmente con ellas, ya que al conocer un poco sobre el tema podemos razonar mejor al momento de plantear un objetivo y las posibles problemáticas.  

Algunos de los beneficios que podemos obtener al trabajar con esta base, como ya mencionamos es la accesibilidad de manejar las variables y los datos proporcionados, los géneros pertenecientes a las películas/series están muy detallados y nos gustó que se proporcionará la cantidad de votos obtenidos.

**5) Objetivo final**

Objetivo principal:  
+ Predecir el rating de una película/serie agregada al catálogo recientemente

**6) Mencionar preguntas de interés**
+ ¿Cuáles son los géneros preferidos?
+ ¿El rating de las películas/series se ve afectado por la duración?

## Limpieza de datos de la base principal: "MOVIES DATASET FOR FEATURE EXTRACION, PREDICTION"

Empezaremos con la importación de librerías, leyendo nuestro archivo y con una visualización no completa de nuestra base.

In [19]:
import pandas as pd
import numpy as np

base1 = pd.read_csv('movies.csv', sep = ',') #Leyendo el archivo csv  
base1

MOVIES         YEAR  \
0                           Blood Red Sky        -2021   
1     Masters of the Universe: Revelation     (2021– )   
2                        The Walking Dead  (2010–2022)   
3                          Rick and Morty     (2013– )   
4                         Army of Thieves        -2021   
...                                   ...          ...   
9994                       The Imperfects     (2021– )   
9995                               Arcane     (2021– )   
9996                    Heart of Invictus     (2022– )   
9997                       The Imperfects     (2021– )   
9998                       The Imperfects     (2021– )   

                                           GENRE  RATING  \
0         \nAction, Horror, Thriller                 6.1   
1     \nAnimation, Action, Adventure                 5.0   
2          \nDrama, Horror, Thriller                 8.2   
3     \nAnimation, Adventure, Comedy                 9.2   
4            \nAction, Crime, Horror                 NaN   
...                                          ...     ...   
9994     \nAdventure, Drama, Fantasy                 NaN   
9995  \nAnimation, Action, Adventure                 NaN   
9996            \nDocumentary, Sport                 NaN   
9997     \nAdventure, Drama, Fantasy                 NaN   
9998     \nAdventure, Drama, Fantasy                 NaN   

                                               ONE-LINE  \
0     \nA woman with a mysterious illness is forced ...   
1     \nThe war for Eternia begins again in what may...   
2     \nSheriff Deputy Rick Grimes wakes up from a c...   
3     \nAn animated series that follows the exploits...   
4     \nA prequel, set before the events of Army of ...   
...                                                 ...   
9994                                     \nAdd a Plot\n   
9995                                     \nAdd a Plot\n   
9996                                     \nAdd a Plot\n   
9997                                     \nAdd a Plot\n   
9998                                     \nAdd a Plot\n   

                                                  STARS     VOTES  RunTime  \
0     \n    Director:\nPeter Thorwarth\n| \n    Star...   21062.0    121.0   
1     \n            \n    Stars:\nChris Wood, \nSara...   17870.0     25.0   
2     \n            \n    Stars:\nAndrew Lincoln, \n...  885805.0     44.0   
3     \n            \n    Stars:\nJustin Roiland, \n...  414849.0     23.0   
4     \n    Director:\nMatthias Schweighöfer\n| \n  ...       NaN      NaN   
...                                                 ...       ...      ...   
9994  \n            \n    Stars:\nMorgan Taylor Camp...       NaN      NaN   
9995                                                 \n       NaN      NaN   
9996  \n    Director:\nOrlando von Einsiedel\n| \n  ...       NaN      NaN   
9997  \n    Director:\nJovanka Vuckovic\n| \n    Sta...       NaN      NaN   
9998  \n    Director:\nJovanka Vuckovic\n| \n    Sta...       NaN      NaN   

     Gross  
0      NaN  
1      NaN  
2      NaN  
3      NaN  
4      NaN  
...    ...  
9994   NaN  
9995   NaN  
9996   NaN  
9997   NaN  
9998   NaN  

[9999 rows x 9 columns]

## Visualización

A continuación verificaremos cuántas filas y columnas tiene nuestra base, esto para darnos más idea de la dimensión que tiene.

In [20]:
base1.shape #Dimensión de la base

(9999, 9)

Nuestra base tiene 9999 filas y 9 columnas 

Ahora con la función ".columns" podremos ver el nombre de las 9 columnas.

Vemos el nombre de las columnas e identificamos los nombres que podemos cambiar, ya sea para que tengan un formato más uniforme o porque no lo creemos adecuado o se nos va a dificultar trabajar con ese nombre.

In [21]:
base1.columns #Nombres de columnas de la base

Index(['MOVIES', 'YEAR', 'GENRE', 'RATING', 'ONE-LINE', 'STARS', 'VOTES',
       'RunTime', 'Gross'],
      dtype='object')

<div style="text-align: justify"> Ahora haremos una base nueva con los nombres de las columnas actualizados.   
<div style="text-align: justify">
Cambiamos el nombre de las columnas para que todos estuvieran en mayusculas, la columna de MOVIES se nos hizo más práctico llamarla NAME ya que ahí va el nombre de la pelicula o serie, tambien RunTime, que nos explica la duración, lo cambiamos solo por TIME para que fuera más corto y en mayúsculas, ONE-LINE lo cambiamos para no batallar con el guión y creímos más entendible que fuera DESCRIPTION ya que de eso se trata la columna, nos muestra la descripción de la película/serie.</div>

In [22]:
base2 = base1.rename(columns = {'MOVIES':'NAME','ONE-LINE':'DESCRIPTION','RunTime':'TIME','Gross':'GROSS'}) #renombrando las columnas
base2

NAME         YEAR  \
0                           Blood Red Sky        -2021   
1     Masters of the Universe: Revelation     (2021– )   
2                        The Walking Dead  (2010–2022)   
3                          Rick and Morty     (2013– )   
4                         Army of Thieves        -2021   
...                                   ...          ...   
9994                       The Imperfects     (2021– )   
9995                               Arcane     (2021– )   
9996                    Heart of Invictus     (2022– )   
9997                       The Imperfects     (2021– )   
9998                       The Imperfects     (2021– )   

                                           GENRE  RATING  \
0         \nAction, Horror, Thriller                 6.1   
1     \nAnimation, Action, Adventure                 5.0   
2          \nDrama, Horror, Thriller                 8.2   
3     \nAnimation, Adventure, Comedy                 9.2   
4            \nAction, Crime, Horror                 NaN   
...                                          ...     ...   
9994     \nAdventure, Drama, Fantasy                 NaN   
9995  \nAnimation, Action, Adventure                 NaN   
9996            \nDocumentary, Sport                 NaN   
9997     \nAdventure, Drama, Fantasy                 NaN   
9998     \nAdventure, Drama, Fantasy                 NaN   

                                            DESCRIPTION  \
0     \nA woman with a mysterious illness is forced ...   
1     \nThe war for Eternia begins again in what may...   
2     \nSheriff Deputy Rick Grimes wakes up from a c...   
3     \nAn animated series that follows the exploits...   
4     \nA prequel, set before the events of Army of ...   
...                                                 ...   
9994                                     \nAdd a Plot\n   
9995                                     \nAdd a Plot\n   
9996                                     \nAdd a Plot\n   
9997                                     \nAdd a Plot\n   
9998                                     \nAdd a Plot\n   

                                                  STARS     VOTES   TIME GROSS  
0     \n    Director:\nPeter Thorwarth\n| \n    Star...   21062.0  121.0   NaN  
1     \n            \n    Stars:\nChris Wood, \nSara...   17870.0   25.0   NaN  
2     \n            \n    Stars:\nAndrew Lincoln, \n...  885805.0   44.0   NaN  
3     \n            \n    Stars:\nJustin Roiland, \n...  414849.0   23.0   NaN  
4     \n    Director:\nMatthias Schweighöfer\n| \n  ...       NaN    NaN   NaN  
...                                                 ...       ...    ...   ...  
9994  \n            \n    Stars:\nMorgan Taylor Camp...       NaN    NaN   NaN  
9995                                                 \n       NaN    NaN   NaN  
9996  \n    Director:\nOrlando von Einsiedel\n| \n  ...       NaN    NaN   NaN  
9997  \n    Director:\nJovanka Vuckovic\n| \n    Sta...       NaN    NaN   NaN  
9998  \n    Director:\nJovanka Vuckovic\n| \n    Sta...       NaN    NaN   NaN  

[9999 rows x 9 columns]

Después con la función ".dtypes" veremos los tipos de datos en cada columna esto para ver si tendremos que cambiar algún tipo o si los creemos adecuados.

In [23]:
base2.dtypes #Tipos de datos en cada columna

NAME            object
YEAR            object
GENRE           object
RATING         float64
DESCRIPTION     object
STARS           object
VOTES          float64
TIME           float64
GROSS           object
dtype: object

Observamos que nuestros valores numéricos son: el rating, los votos y la duración; lo cual es correcto.

## Valores nulos

In [24]:
base2.isnull().any() #nos dice si hay datos nulos en cada columna

NAME           False
YEAR            True
GENRE           True
RATING          True
DESCRIPTION    False
STARS          False
VOTES           True
TIME            True
GROSS           True
dtype: bool

Vemos que tenemos valores nulos en las columnas de YEAR, GENRE, RATING, VOTES, TIME y GROSS. A continuación veremos cuantos datos nulos hay en cada una

In [25]:
base2.isnull().sum() #Cuantos datos nulos tenemos por columna

NAME              0
YEAR            644
GENRE            80
RATING         1820
DESCRIPTION       0
STARS             0
VOTES          1820
TIME           2958
GROSS          9539
dtype: int64

Al ver la gran cantidad de valores nulos, decidimos analizar si todas las columnas nos interesan, en caso de no ser tan relevantes procederemos a borrarlas.

<div style="text-align: justify">A simple vista la columna "GROSS" que nos muestra el total del dinero ganado en todo el mundo en millones de dólares, no nos está dando mucha información porque de 9999 películas/series que tenemos, no nos está dando información de 9539 así que no podremos trabajar con tan pocos datos y debido a que no analizaremos esa parte monetaria de la base de datos podemos quitarla. Igualmente las columnas "DESCRIPTION" y "STARS" serán eliminadas; ya que la primera, no nos resulta útil porque como cada película o serie tiene su propia descripción no podriamos compararlas entre ellas y la última porque nuestra investigación no va a tomar en cuenta a los actores participantes de cada cinta.</div>

## Eliminación de columnas

In [26]:
base3 = base2.drop(['GROSS', 'DESCRIPTION', 'STARS'], axis = 1 ) #Eliminando las columnas

In [27]:
base3

NAME         YEAR  \
0                           Blood Red Sky        -2021   
1     Masters of the Universe: Revelation     (2021– )   
2                        The Walking Dead  (2010–2022)   
3                          Rick and Morty     (2013– )   
4                         Army of Thieves        -2021   
...                                   ...          ...   
9994                       The Imperfects     (2021– )   
9995                               Arcane     (2021– )   
9996                    Heart of Invictus     (2022– )   
9997                       The Imperfects     (2021– )   
9998                       The Imperfects     (2021– )   

                                           GENRE  RATING     VOTES   TIME  
0         \nAction, Horror, Thriller                 6.1   21062.0  121.0  
1     \nAnimation, Action, Adventure                 5.0   17870.0   25.0  
2          \nDrama, Horror, Thriller                 8.2  885805.0   44.0  
3     \nAnimation, Adventure, Comedy                 9.2  414849.0   23.0  
4            \nAction, Crime, Horror                 NaN       NaN    NaN  
...                                          ...     ...       ...    ...  
9994     \nAdventure, Drama, Fantasy                 NaN       NaN    NaN  
9995  \nAnimation, Action, Adventure                 NaN       NaN    NaN  
9996            \nDocumentary, Sport                 NaN       NaN    NaN  
9997     \nAdventure, Drama, Fantasy                 NaN       NaN    NaN  
9998     \nAdventure, Drama, Fantasy                 NaN       NaN    NaN  

[9999 rows x 6 columns]

Podemos ver que las columnas no deseadas se han eliminado correctamente.

Una vez que tenemos solamente las columnas deseadas, nos preguntamos: ¿Qué haremos con los datos nulos restantes?. Para responder esto, hay que estudiar los datos que tenemos. De nueva cuenta revisamos los datos nulos que tenemos.

In [28]:
base3.isnull().sum()

NAME         0
YEAR       644
GENRE       80
RATING    1820
VOTES     1820
TIME      2958
dtype: int64

Como todas las variables restantes son importantes para nuestro análisis, eliminaremos todos aquellos valores nulos que se encuentren todavía y revisaremos la cantidad de datos con la que nos quedaremos

In [29]:
df = base3.dropna() # elimina los elementos nulos
df

NAME         YEAR  \
0                           Blood Red Sky        -2021   
1     Masters of the Universe: Revelation     (2021– )   
2                        The Walking Dead  (2010–2022)   
3                          Rick and Morty     (2013– )   
5                             Outer Banks     (2020– )   
...                                   ...          ...   
9959                               Sexify     (2021– )   
9960                               Sexify     (2021– )   
9961                               Sexify     (2021– )   
9962                               Sexify     (2021– )   
9963                          Castlevania  (2017–2021)   

                                           GENRE  RATING     VOTES   TIME  
0         \nAction, Horror, Thriller                 6.1   21062.0  121.0  
1     \nAnimation, Action, Adventure                 5.0   17870.0   25.0  
2          \nDrama, Horror, Thriller                 8.2  885805.0   44.0  
3     \nAnimation, Adventure, Comedy                 9.2  414849.0   23.0  
5             \nAction, Crime, Drama                 7.6   25858.0   50.0  
...                                          ...     ...       ...    ...  
9959                 \nComedy, Drama                 6.8     108.0   38.0  
9960                 \nComedy, Drama                 7.1     103.0   37.0  
9961                 \nComedy, Drama                 6.9      96.0   40.0  
9962                 \nComedy, Drama                 7.1     100.0   47.0  
9963  \nAnimation, Action, Adventure                 8.2     911.0   24.0  

[6779 rows x 6 columns]

In [30]:
df.shape

(6779, 6)

Aún contamos con 6,779 datos, lo que nos parece una buena cantidad para la investigación.

## Guardar un nuevo archivo

Finalmente, procederemos a guardar todos los cambios que hemos hecho a la base de datos creando un archivo con la base resultante.

In [31]:
df.to_csv('Series_nuevo.csv', index=False, sep = ',')